sdk/core/azure-core/inc/http/http.hpp

```diff
@@ -63,6 +66,191 @@ enum class HttpMethod
+  std::map<std::string, std::string> _retryQueryParameters;
+  // Request can contain no body, or either of next bodies (_bodyBuffer plus size or bodyStream)
+  BodyStream* _bodyStream;
+  BodyBuffer* _bodyBuffer;
+
+  // flag to know where to insert header
+  bool _retryModeEnabled;
+
+  // returns left map plus all items in right
+  // when duplicates, left items are preferred
+  static std::map<std::string, std::string> mergeMaps(
+      std::map<std::string, std::string> const& left,
+      std::map<std::string, std::string> const& right)
+  {
+    auto result = std::map<std::string, std::string>(left);
+    std::for_each(right.begin(), right.end(), [&](auto const& pair) { result.insert(pair); });
+    return result;
+  }
+
+  /**
+   * Create an stream from source string. Then use getline and separator to get as many tokens as
+   * possible and insert each one to result vector
+   */
+  static std::vector<std::string> split(std::string source, char separator)
+  {
+    auto result = std::vector<std::string>();
+
+    auto stringAsStream = std::stringstream(source);
+    auto token = std::string();
+
+    while (std::getline(stringAsStream, token, separator))
+    {
+      result.push_back(token);
+    }
+    return result;
+  }
+
+  /**
+   * Will check if there are any query parameter in url looking for symbol '?'
+   * If it is found, it will insert query parameters to _queryParameters internal field
+   * and remove it from url
+   */
+  std::string parseUrl(std::string url)
+  {
+    auto position = url.find('?');
+
+    if (position == std::string::npos)
+    {
+      return url; // no query parameters. Nothing else to do
+    }
+
+    // Get query parameters string and update
+    auto queryParameters = url.substr(position + 1);
+    url = url.substr(0, position);
+
+    // Split all query parameters (should be separated by &)
+    auto queryParametersVector = Request::split(queryParameters, '&');
+
+    // insert each query parameter to internal field
+    std::for_each(queryParametersVector.begin(), queryParametersVector.end(), [&](auto query) {
+      auto parameter = split(query, '=');
+      // Will throw if parameter in query is not valid (i.e. arg:1)
+      _queryParameters.insert(std::pair<std::string, std::string>(parameter[0], parameter[1]));
+    });
+
+    return url;
+  }
+
+  Request(
+      HttpMethod httpMethod,
+      std::string const& url,
+      BodyStream* bodyStream,
+      BodyBuffer* bodyBuffer)
+      : _method(std::move(httpMethod)), _url(parseUrl(std::move(url))), _bodyStream(bodyStream),
+        _bodyBuffer(bodyBuffer), _retryModeEnabled(false)
+  {
+    // TODO: parse url
+  }
+
+public:
+  Request(Request const&) = default;
+  Request& operator=(Request const&) = default;
+
+  Request(HttpMethod httpMethod, std::string const& url)
+      : Request(httpMethod, url, BodyStream::null, BodyBuffer::null)
+  {
+  }
+
+  Request(HttpMethod httpMethod, std::string const& url, BodyBuffer* bodyBuffer)
```

Comment:
Probably makes sense to pass `BodyBuffer&` rather than by pointer.

Reply:
it was a ref initially
But since we init this with NULL, it was looking weird and got feedback about reference are not typically assigned to null.
we were doing something like 
```c
BodyBuffer null = *(BodyBuffer*)NULL;
````

Reply:
I see, if it needs to be nullable than pointers are indeed okay.